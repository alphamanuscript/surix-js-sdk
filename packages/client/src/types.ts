import { WrappedEntity, WrappedFile } from '@surix/data-helpers';

export interface Project {
  entities: ProjectEntities;
  files: ProjectFiles;
  tags: ProjectTags;
}

export interface ProjectEntities {
  get(id: string): Promise<WrappedEntity>;
  query(query?: Query): Promise<WrappedEntity[]>;
  create(entity: Entity): Promise<WrappedEntity>;
  put(entity: Entity): Promise<WrappedEntity>;
}

export interface ProjectFiles {
  get(id: string): Promise<WrappedFile>;
  list(): Promise<WrappedFile[]>;
}

export interface ProjectTags {
  list(): Promise<TagList>;
}

export interface Query {
  tags?: string[];
  skip?: number;
  limit?: number;
  query?: object;
}

export interface TagListItem {
  name: string;
}

export type TagList = TagListItem[];

export enum FieldType {
  TEXT = 'text',
  BOOLEAN = 'boolean',
  NUMBER = 'number',
  DATETIME = 'datetime',
  PHONE = 'phone',
  EMAIL = 'email',
  FILE = 'file',
  OBJECT = 'object',
  LIST = 'list'
}

export interface FieldPair {
  type: FieldType;
  value: RawEntityData | string | RawEntityData[] | number;
}

export interface RawEntityField extends FieldPair {
}

export interface RawEntityData {
  [fieldName: string]: RawEntityField;
}

/**
 * this interface describe the shape of the entity
 * as used within the system and the API
 */
export interface RawEntity {
  data: RawEntityData;
  tags?: string[];
}

export interface DataField {
  [fieldName: string]: string | number | object | boolean;
}
export interface Entity {
  data: DataField;
  tags?: string[];
  _id?: string;
}