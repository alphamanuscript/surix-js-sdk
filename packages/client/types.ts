import { EntityWrapper } from '@surix/data-helpers';

export interface Project {
  db: ProjectDatabase;
  files: ProjectFiles;
}

export interface ProjectDatabase {
  get(id: string): Promise<EntityWrapper>;
  query(query: Query): Promise<EntityWrapper[]>;
  getTags(): Promise<TagList[]>;
}

export interface ProjectFiles {
  get(id: string): Promise<ApiFile>;
  list(): Promise<ApiFile[]>;
}

export interface Query {
  tags?: string[];
  skip?: number;
  limit?: number;
  query?: object;
}

export interface ApiFile {
  name: string;
  status: 'failed' | 'ready' | 'pending';
  public: boolean;
  downloadUrl: string;
  tags: TagList[];
}

export interface TagListItem {
  name: string;
}

export type TagList = TagListItem[];