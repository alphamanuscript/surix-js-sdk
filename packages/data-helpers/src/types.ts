export class ApiEntity {
    _id: string;
    createdAt: string;
    updatedAt: string;
    data: ObjectValue;
    createdBy: AuthContext;
    tags: string[];
}

export type FieldPathArray = Array<string|number>;

export type FieldType = 'text' | 'boolean' | 'number' | 'datetime'
    | 'file' | 'object' | 'list';

export type FieldValue = string | boolean | FileValue | ListValue | ObjectValue;

export interface FileValue {
    ref: string;
    mimeType: string;
    downloadUrl: string;
}

export type ListValue = ListItemValue[];

export interface ListItemValue {
    type: FieldType;
    value: FieldValue;
}

export interface ObjectValue {
    [fieldName: string]: Field;
}

export interface Field extends TypeValuePair {
    name: string;
    label: string;
}

export interface TypeValuePair {
    type: FieldType;
    value: FieldValue;
}

export interface AuthContext {
    _id: string;
    type: 'user' | 'app' | 'localapp';
}