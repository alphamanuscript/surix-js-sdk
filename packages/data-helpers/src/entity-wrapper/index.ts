export { EntityWrapper, wrapEntity, wrapEntityArray } from './entity-wrapper';
