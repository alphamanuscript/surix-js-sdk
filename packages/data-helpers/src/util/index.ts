export {
  deflateValue,
  makeField,
  toFieldPathArray,
  walkEntityPath,
  walkFieldPath,
  expandQuery
} from './util';