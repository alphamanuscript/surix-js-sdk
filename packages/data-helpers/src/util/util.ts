import {
  ApiEntity,
  Field,
  FieldPathArray,
  FieldType,
  FieldValue,
  ListValue,
  ObjectValue, 
  TypeValuePair} from '../types';

type WalkFieldArg = TypeValuePair | Field;
type WalkFieldReturn = TypeValuePair | Field | undefined;

export function toFieldPathArray (fieldPath: string): FieldPathArray {
  return fieldPath ? fieldPath.split('.') : [];
}

export function walkEntityPath (entity: ApiEntity, path: FieldPathArray): WalkFieldReturn {
  const [firstKey, ...pathRemaining] = path;
  if (firstKey in entity.data) {
    return walkFieldPath(entity.data[firstKey], pathRemaining);
  }
}

export function walkFieldPath (field: WalkFieldArg, path: FieldPathArray): WalkFieldReturn {
  const [firstKey, ...restPath] = path;
  if (typeof firstKey === 'undefined') {
    return field;
  }
  if (field.type === 'object' || field.type === 'list') {
    const value = field.value as ObjectValue | ListValue;
    if (firstKey in value) {
      const subField = value[firstKey];
      return walkFieldPath(subField, restPath);
    }
  }
}

export function deflateValue (value: any, type: FieldType): any {
  switch (type) {
    case 'datetime':
      return new Date(value);
    case 'number':
      return Number(value);
    case 'list':
      return deflateList(value);
    case 'object':
      return deflateObject(value);
    case 'boolean':
    case 'text':
    case 'file':
    default:
      return value;
  }
}

export function deflateText (textValue: string): string {
  return textValue;
}

export function deflateNumber (numValue: string): number {
  return Number(numValue);
}

export function deflateObject (obj: ObjectValue): object {
  return Object.keys(obj)
    .reduce((result, key) => {
      const { type, value } = obj[key];
      return {
        ...result,
        [key]: deflateValue(value, type)
      };
    }, {});
}

export function deflateList (list: ListValue): any[] {
  return list.map(({ type, value }) => deflateValue(value, type));
}

export function makeField (value: FieldValue, type: FieldType, name: string, label: string): Field {
  return {
    value,
    type,
    name,
    label
  };
}