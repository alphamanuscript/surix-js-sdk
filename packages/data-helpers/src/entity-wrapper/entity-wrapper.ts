import { ApiEntity, AuthContext, Field, FieldPathArray, FieldType, TypeValuePair } from '../types';
import { deflateValue, toFieldPathArray, walkEntityPath } from '../util';

type FieldOrValuePair = Field | TypeValuePair;
type NullableFieldOrValuePair = FieldOrValuePair | undefined;

const normalizeFieldPath = (fieldPath: string|FieldPathArray): FieldPathArray =>
   Array.isArray(fieldPath) && fieldPath || toFieldPathArray(fieldPath);

export class EntityWrapper {
  private _entity: ApiEntity;
  private _dataCache: object;

  /**
   * date when the entity was created
   */
  createdAt: Date;

  /**
   * date when the entity was last updated
   */
  updatedAt: Date;

  /**
   * 
   * @param rawEntity raw entity from the Surix API
   */
  constructor (rawEntity: ApiEntity) {
    this._entity = rawEntity;
    this.createdAt = new Date(this._entity.createdAt);
    this.updatedAt = new Date(this._entity.updatedAt);
  }

  /**
   * the raw entity from the Surix API
   */
  get rawEntity (): ApiEntity {
    return this._entity;
  }

  /**
   * the unique id of the entity
   */
  get _id (): string {
    return this._entity._id;
  }

  /**
   * the unique id of the entity
   * @alias _id
   */
  get id (): string {
    return this._id;
  }

  /**
   * the list of tags that the entity belongs to
   */
  get tags (): string[] {
    return this._entity.tags;
  }

  /**
   * the principal that created the entity
   */
  get createdBy (): AuthContext {
    return this._entity.createdBy;
  }

  /**
   * returns the plain value of the field at the specified path
   * @param fieldPath the key of the field, this can be a
   * dotted string "field.nestedField" or an array ["field", "nestedField"]
   * the path can include field names as well as array indices
   * @param defaultValue 
   */
  get (fieldPath: string | FieldPathArray, defaultValue?: any): any {
    const pathArray = normalizeFieldPath(fieldPath);
    return this._value(pathArray, defaultValue);
  }

  private _value (fieldPath: FieldPathArray, defaultValue?: any): any {
    const field = walkEntityPath(this._entity, fieldPath);
    return field ? deflateValue(field.value, field.type) : defaultValue;
  }

  /**
   * returns the type of the field with the specified key
   * @param fieldPath the key of the field, this can be a
   * dotted string "field.nestedField" or an array ["field", "nestedField"]
   * the path can include field names as well as array indices
   * @param defaultType the type to return if the entity does not have the field
   */
  type (fieldPath: string | FieldPathArray, defaultType?: FieldType): FieldType|undefined {
    const pathArray = normalizeFieldPath(fieldPath);
    const field = walkEntityPath(this._entity, pathArray);
    return field ? field.type : defaultType;
  }

  /**
   * returns the label of the field with the specified key
   * @param fieldPath the key of the field, this can be a
   * dotted string "field.nestedField" or an array ["field", "nestedField"]
   * the path can include field names as well as array indices
   * @param defaultLabel the label to return if the field does not exist
   */
  label (fieldPath: string | FieldPathArray, defaultLabel?: string): string|undefined {
    const pathArray = normalizeFieldPath(fieldPath);
    const field = walkEntityPath(this._entity, pathArray) as Field;
    return field && field.label ? field.label : defaultLabel;
  } 

  /**
   * returns the raw field at the specified key
   * @param fieldPath the key of the field, this can be a
   * dotted string "field.nestedField" or an array ["field", "nestedField"]
   * the path can include field names as well as array indices
   * @param defaultField the type to return if the entity does not have the field
   */
  field (fieldPath: string | FieldPathArray, defaultField?: FieldOrValuePair):
  NullableFieldOrValuePair {
    const pathArray = normalizeFieldPath(fieldPath);
    const field = walkEntityPath(this._entity, pathArray);
    return field || defaultField;
  } 

  /**
   * returns the entity data
   * as plain (deflated) key-value pair
   */
  data (): object {
    if (!this._dataCache) {
      this._dataCache = deflateValue(this._entity.data, 'object');
    }
    return this._dataCache;
  }
}

export function wrapEntity (entity: ApiEntity): EntityWrapper {
  return new EntityWrapper(entity);
}

export function wrapEntityArray (entities: ApiEntity[]): EntityWrapper[] {
  return entities.map(wrapEntity);
}