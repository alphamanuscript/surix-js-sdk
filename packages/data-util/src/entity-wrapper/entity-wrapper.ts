import { ApiEntity, AuthContext, FieldPathArray } from '../types';
import { dehydrateValue, toFieldPathArray, walkEntityPath } from '../util';

export class EntityWrapper {
  private _entity: ApiEntity;
  constructor (rawEntity: ApiEntity) {
    this._entity = rawEntity;
  }

  get rawEntity (): ApiEntity {
    return this._entity;
  }

  get _id (): string {
    return this._entity._id;
  }

  get tags (): string[] {
    return this._entity.tags;
  }

  get createdBy (): AuthContext {
    return this._entity.createdBy;
  }

  get createdAt (): Date {
    return new Date(this._entity.createdAt);
  }

  get updatedAt (): Date {
    return new Date(this._entity.updatedAt);
  }

  value (fieldPath: string | FieldPathArray): any {
    const pathArray = Array.isArray(fieldPath) && fieldPath || toFieldPathArray(fieldPath);
    return this._value(pathArray);
  }
  _value (fieldPath: FieldPathArray): any {
    const field = walkEntityPath(this._entity, fieldPath);
    if (field) {
      return dehydrateValue(field.value, field.type);
    }
  }
    
}