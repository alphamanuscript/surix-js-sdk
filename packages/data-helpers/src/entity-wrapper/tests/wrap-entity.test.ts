import { ApiEntity } from '../../types';
import { wrapEntity, wrapEntityArray } from '../entity-wrapper';

import * as util from '../../util';

describe('wrapEntity', () => {
  it('should transform entity into an EntityWrapper', () => {
    const entity: ApiEntity = {
      _id: 'ent1',
      data: {
        name: util.makeField('Foo', 'text')
      },
      createdBy: {
        _id: 'u1',
        type: 'user'
      },
      updatedBy: {
        _id: 'u2',
        type: 'user'
      },
      createdAt: '2017-02-22T09:19:33.885Z',
      updatedAt: '2018-10-22T09:19:33.885Z',
      tags: ['t1', 't2']
    };
    const wrapped = wrapEntity(entity);
    expect(wrapped).toMatchSnapshot();
  });
});

describe('wrapEntityArray', () => {
  it('should transform an array of entities into an array of EntityWrapper', () => {
    const entities: ApiEntity[] = [
      {
        _id: 'ent1',
        data: {
          name: util.makeField('Foo', 'text')
        },
        createdBy: {
          _id: 'u1',
          type: 'user'
        },
        updatedBy: {
          _id: 'u2',
          type: 'user'
        },
        createdAt: '2017-02-22T09:19:33.885Z',
        updatedAt: '2018-10-22T09:19:33.885Z',
        tags: ['t1', 't2']
      },
      {
        _id: 'ent2',
        data: {
          name: util.makeField('Bar', 'text')
        },
        createdBy: {
          _id: 'u2',
          type: 'user'
        },
        updatedBy: {
          _id: 'u3',
          type: 'user'
        },
        createdAt: '2017-02-23T09:19:33.885Z',
        updatedAt: '2018-10-23T09:19:33.885Z',
        tags: ['t1', 't2']
      }
    ];
    const wrapped = wrapEntityArray(entities);
    expect(wrapped).toMatchSnapshot();
  });
});