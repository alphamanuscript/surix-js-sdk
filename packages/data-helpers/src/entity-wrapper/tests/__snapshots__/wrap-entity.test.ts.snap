// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`wrapEntity should transform entity into an EntityWrapper 1`] = `
EntityWrapper {
  "_entity": Object {
    "_id": "ent1",
    "createdAt": "2017-02-22T09:19:33.885Z",
    "createdBy": Object {
      "_id": "u1",
      "type": "user",
    },
    "data": Object {
      "name": Object {
        "label": "Name",
        "name": "name",
        "type": "text",
        "value": "Foo",
      },
    },
    "tags": Array [
      "t1",
      "t2",
    ],
    "updatedAt": "2018-10-22T09:19:33.885Z",
  },
  "createdAt": 2017-02-22T09:19:33.885Z,
  "updatedAt": 2018-10-22T09:19:33.885Z,
}
`;

exports[`wrapEntityArray should transform an array of entities into an array of EntityWrapper 1`] = `
Array [
  EntityWrapper {
    "_entity": Object {
      "_id": "ent1",
      "createdAt": "2017-02-22T09:19:33.885Z",
      "createdBy": Object {
        "_id": "u1",
        "type": "user",
      },
      "data": Object {
        "name": Object {
          "label": "Name",
          "name": "name",
          "type": "text",
          "value": "Foo",
        },
      },
      "tags": Array [
        "t1",
        "t2",
      ],
      "updatedAt": "2018-10-22T09:19:33.885Z",
    },
    "createdAt": 2017-02-22T09:19:33.885Z,
    "updatedAt": 2018-10-22T09:19:33.885Z,
  },
  EntityWrapper {
    "_entity": Object {
      "_id": "ent2",
      "createdAt": "2017-02-23T09:19:33.885Z",
      "createdBy": Object {
        "_id": "u2",
        "type": "user",
      },
      "data": Object {
        "name": Object {
          "label": "Name",
          "name": "name",
          "type": "text",
          "value": "Bar",
        },
      },
      "tags": Array [
        "t1",
        "t2",
      ],
      "updatedAt": "2018-10-23T09:19:33.885Z",
    },
    "createdAt": 2017-02-23T09:19:33.885Z,
    "updatedAt": 2018-10-23T09:19:33.885Z,
  },
]
`;
